### PR TITLE
Surface tracker apriltag support

### PR DESCRIPTION
**Pull request by romanroibu**
##### TODO

- [x] Question: Should there be an option which marker detectors to use? Or should surface tracking just work with both square markers and apriltag markers all the time?

Solution: Use both detectors exclusively to each other, opting to April Tag by default. Users will have the option to switch to the legacy markers if necessary.

- [x] ~~When using both `square` and `apriltag` detectors at the same time, in some cases, some `apriltag` markers are recognized by both detectors. This can be spotted when there are overlapping visualizations. Possible solutions:~~
    - ~~*Do nothing.* Let the user decide how to handle cases when both detectors are enabled, and the same marker is recognized by both. One obvious way is to only enable one detector (and use the corresponding set of tags) at a time.~~
    - ~~*Use the tag with highest confidence.* Detect the overlap (marker centers are "too close") and pick the marker with the highest confidence. An alternative is to just pick the `apriltag` marker in case of an overlap, since `square` don't get recognized by the `apriltag` detector.~~

Edit by @papr

### CODE REVIEW

**Comment by papr** [comment]
I like the idea of both detectors running in parallel without adding further ui options for the user. Regarding conflicting detections: I like the option to always prioritize apriltags. How difficult is it to find overlapping/duplicated detections?

**Comment by romanroibu** [comment]
Not sure yet; will investigate how well can I detect overlaps

**Comment by papr** [comment]
@romanroibu Pleae make sure that this PR does not overwrite any of the fixes in #1555

**Comment by romanroibu** [comment]
@papr 

> - Fixed a bug where the offline surface tracker did not use robust marker detection correctly, which lead to a worse marker detection performance.
> - Also set the minimum marker detection confidence to zero to mirror behavior of previous version of surface tracker.

is there data that I can use to test this?

> - Fixed a bug where the heatmap export had switched horizontal and vertical resolution.
> - Reverted a previous commit that removed mapping of fixations to the surface during export.

not sure how to test these changes; any suggestions?

File: pupil_src/shared_modules/surface_tracker/surface_tracker_offline.py [hunk 1]
```diff
@@ -97,6 +101,20 @@ def ui_info_text(self):
     def supported_heatmap_modes(self):
         return [Heatmap_Mode.WITHIN_SURFACE, Heatmap_Mode.ACROSS_SURFACES]
 
+    def _init_marker_detection_modes(self):
+        # This method should be called after `_init_marker_cache` to ensure that the cache is filled in.
+        assert self.marker_cache is not None
+
+        # Filter out non-filled frames where the cache entry is None.
+        # Chain the remaining entries (which are lists) to get a flat sequence.
+        cached_surface_marker_sequence = itertools.chain(*filter(None, self.marker_cache))
```
[the remark this answers is not in the document]
Comment (papr) [reply pupil_src/shared_modules/surface_tracker/surface_tracker_offline.py]: Also, are you sure that `filter(None, LIST)` filters only entries that are None? In my test it filters everything that does not evaluate to `True`. Maybe `filter(lambda x: x is None, ...)` is more appropriate?

File: pupil_src/shared_modules/surface_tracker/surface_marker_detector.py [hunk 2]
```diff
@@ -0,0 +1,414 @@
...
+    def all_supported_cases() -> typing.Set["Surface_Marker_Detector_Mode"]:
+        return set(Surface_Marker_Detector_Mode)
+
+    @staticmethod
+    def from_marker(marker: Surface_Marker) -> "Surface_Marker_Detector_Mode":
+        marker_type = marker.marker_type
+        if marker_type == Surface_Marker_Type.SQUARE:
+            return Surface_Marker_Detector_Mode.SQUARE_MARKER
+        if marker_type == Surface_Marker_Type.APRILTAG_V2:
+            return Surface_Marker_Detector_Mode.APRILTAG_MARKER
```
[the remark this answers is not in the document]
Comment (romanroibu) [reply pupil_src/shared_modules/surface_tracker/surface_marker_detector.py]: well, the idea is that `Surface_Marker_Type` is more of an implementation detail for the `surface_tracker` package. For example, it distinguishes not only the type of the marker (e.g. `square` vs. `apriltag`) but also the version of the libraries (e.g. `apriltag_v2` vs. `apriltag_v3`). From an interface point of view, when the user sets the "mode" of the detector to `Surface_Marker_Detector_Mode.APRILTAG_MARKER`, the user gets a list of `Surface_Marker` instances, and shouldn't care what the `marker_type` is. The `marker_type` is only relevant when the marker is serialized/deserialized.

[the remark this answers is not in the document]
Comment (papr) [reply pupil_src/shared_modules/surface_tracker/surface_marker_detector.py]: Understood, thank you.

**Pull request merged**